closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -2354,14 +2362,138 @@ Java_jdk_crypto_jniprovider_NativeCrypto_ECCreatePrivateKey
 
     (*env)->ReleasePrimitiveArrayCritical(env, s, nativeS, JNI_ABORT);
 
+    if (NULL != sBN) {
+        (*OSSL_BN_free)(sBN);
+    }
+
     if (0 == ret) {
         return -1;
     }
 
     return 0;
 }
 
-/* Encode an EC Elliptic Curve over a Prime Field
+/* Encode an EC Elliptic Curve over a Prime Field */
+EC_KEY* ECEncodeGFp(BIGNUM *aBN,
+                    BIGNUM *bBN,
+                    BIGNUM *pBN,
+                    BIGNUM *xBN,
+                    BIGNUM *yBN,
+                    BIGNUM *nBN,
+                    BIGNUM *hBN)
+{
+    EC_KEY *key = NULL;
+    EC_GROUP *group = NULL;
+    EC_POINT *generator = NULL;
+    BN_CTX *ctx = NULL;
+    int ret = 0;
+
+    ctx = (*OSSL_BN_CTX_new)();
+    if (NULL == ctx) {
+        goto cleanup;
+    }
+
+    group = (*OSSL_EC_GROUP_new_curve_GFp)(pBN, aBN, bBN, ctx);
+    if (NULL == group)
+        goto cleanup;
+
+    generator = (*OSSL_EC_POINT_new)(group);
+    if (NULL == generator)
+        goto cleanup;
+
+    ret = (*OSSL_EC_POINT_set_affine_coordinates_GFp)(group, generator, xBN, yBN, ctx);
+    if (0 == ret)
+        goto cleanup;
+
+    ret = (*OSSL_EC_GROUP_set_generator)(group, generator, nBN, hBN);
+    if (0 == ret)
+        goto cleanup;
+
+    key = (*OSSL_EC_KEY_new)();
+    if (NULL == key)
+        goto cleanup;
+
+    ret = (*OSSL_EC_KEY_set_group)(key, group);
+    if (0 == ret) {
+        (*OSSL_EC_KEY_free)(key);
+        key = NULL;
+    }
+
+cleanup:
+    if (NULL != generator)
+        (*OSSL_EC_POINT_free)(generator);
+
+    if (NULL != group)
+        (*OSSL_EC_GROUP_free)(group);
+
+    if (NULL != ctx)
+        (*OSSL_BN_CTX_free)(ctx);
+
+    return key;
+}
+
+/* Encode an EC Elliptic Curve over a Binary Field */
+EC_KEY* ECEncodeGF2m(BIGNUM *aBN,
+                     BIGNUM *bBN,
+                     BIGNUM *pBN,
+                     BIGNUM *xBN,
+                     BIGNUM *yBN,
+                     BIGNUM *nBN,
+                     BIGNUM *hBN)
+{
+    EC_KEY *key = NULL;
+    EC_GROUP *group = NULL;
+    EC_POINT *generator = NULL;
+    BN_CTX *ctx = NULL;
+    int ret = 0;
+
+    if (JNI_FALSE == OSSL_ECGF2M)
+        return NULL;
+
+    ctx = (*OSSL_BN_CTX_new)();
+    if (NULL == ctx)
+        goto cleanup;
+
+    group = (*OSSL_EC_GROUP_new_curve_GF2m)(pBN, aBN, bBN, ctx);
+    if (NULL == group)
+        goto cleanup;
+
+    generator = (*OSSL_EC_POINT_new)(group);
+    if (NULL == generator)
+        goto cleanup;
+
+    ret = (*OSSL_EC_POINT_set_affine_coordinates_GF2m)(group, generator, xBN, yBN, ctx);
+    if (0 == ret)
+        return NULL;
```

Comment:
Shouldn't this be `goto cleanup;`?

Reply:
Good catch here. I agree that we want to `goto cleanup` in order to cleanup any allocated objects, then return `NULL` instead of directly returning `NULL` in this spot.